icarusalg/Utilities/AssnsCrosser.h

```diff
@@ -0,0 +1,1749 @@
+{}
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+template <typename Event>
+icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>::AssnsCrosser(
+  Event const& event,
+  StartSpecs<KeyType> startSpecs,
+  InputSpecs<OtherTypes>... otherInputSpecs
+)
+  : fAssnsMap
+    { prepare(event, std::move(startSpecs), std::move(otherInputSpecs)... ) }
+{}
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+auto icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>::assPtr
+  (KeyPtr_t const& keyPtr) const -> TargetPtr_t const&
+{
+  TargetPtrs_t const& targets = assPtrs(keyPtr);
+  if (targets.size() > 1) {
+    // using LogicError because that's what art::FindOne does
+    throw art::Exception{ art::errors::LogicError }
+      << "AssnsCrosser::assPtr(): there are " << targets.size() << " "
+      << lar::debug::demangle<Target_t>() << " objects associated to Ptr<"
+      << lar::debug::demangle<Key_t>() << ">=" << keyPtr << "!\n";
+  }
+  return targets.empty()? NullTargetPtr: targets.front();
+} // icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>::assPtr()
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+template <typename Event>
+auto icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>::prepare(
+  Event const& event,
+  StartSpecs<KeyType> startSpecs, InputSpecs<OtherTypes>... otherInputSpecs
+) const -> AssnsMap_t
+{
+  std::optional<details::PointerSelector<Key_t>> keySelector
+    = keysFromSpecs(event, startSpecs);
+  HoppingAlgo const algo
+    = chooseTraversalAlgorithm(startSpecs, otherInputSpecs...);
+  switch (algo) {
+    case HoppingAlgo::forward:
+      return details::MapJoiner<KeyType, OtherTypes...>::joinForward
+        (event, std::move(otherInputSpecs)..., keySelector);
+    case HoppingAlgo::backward:
+      return details::MapJoiner<KeyType, OtherTypes...>::joinBackward
+        (event, std::move(otherInputSpecs)... );
+    default:
+      throw std::logic_error
+        { "Unexpected direction: " + std::to_string(static_cast<int>(algo)) };
+  } // switch
+} // icarus::ns::util::AssnsCrosser<>::prepare()
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+auto icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>
+  ::chooseTraversalAlgorithm
+(
+  StartSpecs<KeyType> const& startSpecs,
+  InputSpecs<OtherTypes> const&... otherInputSpecs
+) const -> HoppingAlgo {
+  /*
+   * If there is a start specification, we need (and can) to go forward.
+   * Otherwise, we go backward (faster) unless there is no specification for
+   * the last hop (in which case we can't start from the back).
+   * 
+   * When both algorithms are available, we prefer the backward one.
```

Comment:
> I was thinking that the code would prefer to choose the backward one

That is in fact the case.
If there is a special start specification (i.e. if you want to tell the algorithm exactly which pointers you are going to query about), the forward algorithm is mandatory.
In all other cases, either the backward one is chosen (even in case of a fully specified chain where for sure both algorithms work), or a fatal error is emitted.